File: src/intervalwidget.py
import time

# pylint: disable=wrong-import-position
import cairo
import gi
gi.require_version('GLib', '2.0')
from gi.repository import GLib  # noqa: E402

from config import Config  # noqa: E402
from sessionview import SessionView  # noqa: E402
from utils import format_mm_ss  # noqa: E402
# pylint: enable=wrong-import-position


class IntervalWidget(SessionView):
    def __init__(self, config: Config, feed_url: str):
        super().__init__(config, feed_url)
        self.fontoptions = cairo.FontOptions()
        self.fontoptions.set_antialias(cairo.ANTIALIAS_SUBPIXEL)

        self.intervals = []
        self.playing = False
        self.start_time = None
        self.current_interval_index = None
        self.pause_time = None
        self.elapsed = None
        self.connect("draw", self.on_draw)

    def play(self, video_id):
        self.intervals = self.read_intervals(video_id)
        self.start_time = time.monotonic()
        self.current_interval_index = 0
        self.playing = True
        GLib.timeout_add_seconds(1, self.force_redraw)

    def play_pause(self):
        now = time.monotonic()
        if self.playing:
            self.pause_time = now
            self.elapsed = now - self.start_time
        else:
            self.start_time = now - self.elapsed
        self.playing = not self.playing
        self.force_redraw()

    def force_redraw(self):
        if self.current_interval_index is None:
            return False  # don't call again until restarted
        self.queue_draw()
        return True

    def on_draw(self, drawingarea, context: cairo.Context):
        if self.current_interval_index is None:
            return

        now = time.monotonic() if self.playing else self.pause_time

        while (self.current_interval_index < len(self.intervals)
               and now >= (self.start_time
                           + self.intervals[self.current_interval_index].start_offset
                           + self.intervals[self.current_interval_index].duration)):
            self.current_interval_index += 1

        text_h = 30
        block_h = text_h * 4

        y0 = end_y = 0
        one_minute_h = drawingarea.get_allocated_height() - block_h

        context.select_font_face('Sans', cairo.FontSlant.NORMAL, cairo.FontWeight.NORMAL)
        context.set_antialias(cairo.Antialias.SUBPIXEL)
        context.set_font_size(24)  # TODO: units??

        draw_interval_index = self.current_interval_index
        while draw_interval_index < len(self.intervals):
            draw_interval = self.intervals[draw_interval_index]
            interval_start_time = self.start_time + draw_interval.start_offset
            if interval_start_time < now:
                y = y0
            else:
                start_delta = interval_start_time - now
                y = y0 + start_delta / 60 * one_minute_h
            if y > drawingarea.get_allocated_height():
                break

            if draw_interval_index == self.current_interval_index:
                draw_interval_end = (self.start_time
                                     + draw_interval.start_offset
                                     + draw_interval.duration)
                draw_interval_remaining = draw_interval_end - now
            else:
                draw_interval_remaining = draw_interval.duration
            h = draw_interval_remaining / 60.0 * one_minute_h

            context.rectangle(0,
                              y,
                              drawingarea.get_allocated_width(),
                              h)
            context.set_source_rgb(draw_interval.color[0] / 255.0,
                                   draw_interval.color[1] / 255.0,
                                   draw_interval.color[2] / 255.0)
            context.fill_preserve()
            context.set_source_rgb(0.0, 0.0, 0.0)
            context.stroke()

            text_x = 20  # offset the text slightly into the rectangle
            text_y = y + text_h
            context.move_to(text_x, text_y)
            context.show_text(draw_interval.name)
            text_y += text_h

            context.move_to(text_x, text_y)
            context.show_text(draw_interval.effort)
            text_y += text_h

            context.move_to(text_x, text_y)
            context.show_text(f'RPM: {draw_interval.cadence}')
            text_y += text_h

            context.move_to(text_x, text_y)
            context.show_text(format_mm_ss(draw_interval_remaining))
            text_y += text_h

            end_y = y + h

            draw_interval_index += 1

        context.set_source_rgb(0, 0, 0)
        context.rectangle(0, end_y, drawingarea.get_allocated_width(), drawingarea.get_allocated_height() - end_y)
        context.fill_preserve()

File: widgets/intervalwidget.go
package widgets

import (
	"fmt"
	"nsw42/picave/feed"
	"nsw42/picave/profile"
	"nsw42/picave/utils"
	"time"

	"github.com/diamondburned/gotk4/pkg/cairo"
	"github.com/diamondburned/gotk4/pkg/glib/v2"
	"github.com/diamondburned/gotk4/pkg/gtk/v4"
)

type DrawInterval struct {
	feed.IntervalDefinition
	StartOffset time.Duration
	EndOffset   time.Duration
	EffortStr   string
}

type IntervalWidget struct {
	*gtk.DrawingArea
	Profile              *profile.Profile
	Session              *feed.SessionDefinition
	PlayerStartTime      time.Time
	DrawIntervals        []DrawInterval
	CurrentIntervalIndex int
}

const (
	TextXPos        = 8
	TextHeight      = 30
	TextBlockHeight = TextHeight * 4
)

func NewIntervalWidget(profile *profile.Profile) *IntervalWidget {
	rtn := &IntervalWidget{gtk.NewDrawingArea(), profile, nil, time.Time{}, []DrawInterval{}, -1}
	rtn.SetDrawFunc(rtn.OnDraw)
	return rtn
}

func (widget *IntervalWidget) ForceRedraw() bool {
	if widget.CurrentIntervalIndex < 0 {
		return false
	}
	widget.QueueDraw()
	return true
}

func (widget *IntervalWidget) StartPlaying(session *feed.SessionDefinition) {
	widget.Session = session
	widget.DrawIntervals = []DrawInterval{}
	startOffset := time.Duration(0)
	videoFTP := float64(widget.Profile.GetVideoFTPVal(session.VideoId, true))
	videoMaxInt := widget.Profile.GetVideoMaxVal(session.VideoId, true)
	var videoMaxStr string
	if videoMaxInt == 0 {
		videoMaxStr = "MAX"
	} else {
		videoMaxStr = fmt.Sprintf("%d%%", videoMaxInt)
	}
	for _, feedInterval := range session.Intervals {
		var intervalEffort float64
		var intervalEffortStr string
		switch feedInterval.Effort.Type {
		case feed.RelativeToFTPValue:
			intervalEffort = float64(feedInterval.Effort.PercentFTP) * videoFTP / 100.0
			intervalEffortStr = fmt.Sprintf("%dW", int(intervalEffort))
		case feed.MaxEffort:
			intervalEffortStr = videoMaxStr
		}
		widget.DrawIntervals = append(widget.DrawIntervals,
			DrawInterval{
				feedInterval,
				startOffset,
				startOffset + feedInterval.Duration.Duration,
				intervalEffortStr})
		startOffset += feedInterval.Duration.Duration
	}
	widget.CurrentIntervalIndex = 0
	widget.PlayerStartTime = time.Now()
	glib.TimeoutAdd(200, widget.ForceRedraw)
}

func (widget *IntervalWidget) OnDraw(area *gtk.DrawingArea, cr *cairo.Context, width, height int) {
	if widget.Session == nil {
		cr.SetSourceRGB(0, 0, 0)
		cr.Paint()
		return
	}

	now := time.Now()
	intervals := widget.DrawIntervals
	for widget.CurrentIntervalIndex < len(intervals) &&
		now.After(widget.PlayerStartTime.Add(intervals[widget.CurrentIntervalIndex].EndOffset)) {
		widget.CurrentIntervalIndex++
	}

	endY := 0.0
	oneMinuteHeight := float64(area.AllocatedHeight() - TextBlockHeight)
	oneSecondHeight := oneMinuteHeight / 60.

	drawW := float64(area.AllocatedWidth())

	cr.SelectFontFace("Sans", cairo.FontSlantNormal, cairo.FontWeightNormal)
	cr.SetAntialias(cairo.AntialiasDefault)
	cr.SetFontSize(24)

	for drawIndex, drawInterval := range intervals[widget.CurrentIntervalIndex:] {
		var drawIntervalRemaining time.Duration
		var rectY, rectH, textY float64
		if drawIndex == 0 {
			// 0 => current interval, not necessarily the first in the session
			drawIntervalEnd := widget.PlayerStartTime.Add(drawInterval.EndOffset)
			drawIntervalRemaining = drawIntervalEnd.Sub(now)
			rectY = 0
			rectH = drawIntervalRemaining.Seconds() * oneSecondHeight
			textY = min(0, rectH-TextBlockHeight-8) + TextHeight
		} else {
			drawIntervalRemaining = drawInterval.Duration.Duration
			intervalStartTime := widget.PlayerStartTime.Add(drawInterval.StartOffset)
			startDelta := intervalStartTime.Sub(now)
			rectY = startDelta.Seconds() * oneSecondHeight
			rectH = drawIntervalRemaining.Seconds() * oneSecondHeight
			textY = rectY + TextHeight
		}

		cr.Rectangle(0, rectY, drawW, rectH)
		cr.SetSourceRGB(drawInterval.Color.Red, drawInterval.Color.Green, drawInterval.Color.Blue)
		cr.FillPreserve()
		cr.SetSourceRGB(0, 0, 0)
		cr.Stroke()

		cr.MoveTo(TextXPos, textY)
		cr.ShowText(drawInterval.Name)
		textY += TextHeight

		cr.MoveTo(TextXPos, textY)
		cr.ShowText(drawInterval.EffortStr)
		textY += TextHeight

		cr.MoveTo(TextXPos, textY)
		cr.ShowText(fmt.Sprintf("RPM: %d", drawInterval.Cadence))
		textY += TextHeight

		cr.MoveTo(TextXPos, textY)
		cr.ShowText(utils.FormatDurationMMSS(drawIntervalRemaining))
		textY += TextHeight

		endY = rectY + rectH
	}

	cr.SetSourceRGB(0, 0, 0)
	cr.Rectangle(0, endY, drawW, float64(area.AllocatedHeight())-endY)
	cr.FillPreserve()
}
